src/main.cpp

```diff
@@ -6,142 +6,96 @@
-			if(foundDefault){funcName = "default";}
-			else{funcName = "";}
-		}
 
-		if(funcName != ""){ // use the specified function or the default function from the lua script
-			parser[rule.c_str()] = [rule, funcName](const SemanticValues& sv, any&){
-
-				// push input parameters on stack
-				const lua::value params = lua::newtable();
-				params["choice"] = sv.choice();
-				params["column"] = sv.line_info().second;
-				params["line"] = sv.line_info().first;
-				params["matched"] = StringPtr(sv.c_str(), sv.length());
-				params["rule"] = rule;
-
-				const lua::value subnodes = lua::newtable();
-				for (size_t i = 0; i != sv.size(); ++i){
-					subnodes[1 + i] = sv[i].get<lua::value>();
-				}
-				params["subnodes"] = subnodes;
-
-				const lua::value tokens = lua::newtable();
-				for (size_t i = 0; i != sv.tokens.size(); ++i) {
-					tokens[1 + i] = StringPtr(sv.tokens[i].first, sv.tokens[i].second);
-				}
-				params["tokens"] = tokens;
-
-				// find function
-				const lua::value func(lua::globals()[funcName.c_str()]);
-
-				// call lua function and return lua value.
-				return func(params);
-			};
-		}
-		else{ // function not found, no default function in lua script -> just return last matched token
-			parser[rule.c_str()] = [&L](const SemanticValues& sv, any&){
-				if(sv.tokens.size() == 0){
-					lua_pushlstring(L.get(), sv.c_str(), sv.length()); // no tokens, return matched string
-				}
-				else {
-					lua_pushlstring(L.get(), sv.tokens.back().first, sv.tokens.back().second); // return last token
-				}
-
-				// return lua value
-				return lua::value();
-			};
-		}
+	// read action table ( https://stackoverflow.com/questions/6137684/iterate-through-lua-table )
+	lua_pushvalue(L, 2);
+	lua_pushnil(L);
+	while (lua_next(L, -2))
+	{
+		lua_pushvalue(L, -2);
+		string rule = lua_tostring(L, -1); // = key
+		string funcName = ptr2str(lua_topointer(L, -2)); // = value (use the address of the function as its name)
+		lua_pop(L, 1); // pop key
+		lua_setglobal(L, funcName.c_str()); // store function in lua globals (pops from stack)
+		registerReductionRule(pegParser, L, rule, funcName);
 	}
+	lua_pop(L, 1);
+
+
+	// configure parser
+	pegParser->enable_packrat_parsing();
+
 
 	// assign callbacks for parser debugging
 	#ifdef DEBUG_PARSER
-	for(auto r:rules){
+	for(auto& rule:rules){
 
-		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
+		(*pegParser)[rule.c_str()].enter = [rule](const char* s, size_t n, any& dt) {
 			auto& indent = *dt.get<int*>();
-			cout << repeat("|  ", indent) << r << " => \"" << shorten(s, n, DEBUG_STRLEN) << "\"?" << endl;
+			cout << repeat("|  ", indent) << rule << " => \"" << shorten(s, n, DEBUG_STRLEN) << "\"?" << endl;
 			indent++;
 		};
 
-		parser[r.c_str()].leave = [r](const char* s, size_t, size_t matchlen, any& value, any& dt) {
+		(*pegParser)[rule.c_str()].leave = [rule, &L](const char* s, size_t n, size_t matchlen, any& value, any& dt) {
```

Comment:
As above, `&L` here refers to a variable that does not exist anymore when this lambda is called. But `L` is also unused.